woltka/coverage.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2020--, Qiyun Zhu.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+
+"""Functions for calculating coverage of subject sequences.
+
+The "coverage" refers to the ranges of a subject sequence (e.g., genome) that
+are covered by at least one query sequence (e.g., short sequencing read).
+"""
+
+from os import makedirs
+from os.path import join
+from collections import deque
+
+
+def merge_ranges(ranges):
+    """Merge short, fragmental ranges into long, continuous ones.
+
+    Parameters
+    ----------
+    ranges : list of (int, int)
+        Ranges (start, end) to merge.
+
+    Returns
+    -------
+    list of (int, int)
+        Merged ranges.
+    
+    Notes
+    -----
+    Ranges that have overlaps will be merged into one. For example:
+
+    >>> merge_ranges([(1, 3), (2, 4), (6, 8), (7, 9)])
+    [(1, 4), (6, 9)]
+    """
+    res = []
+    res_append = res.append
+    cstart, cend = None, None
+    for start, end in sorted(ranges):
+        if cend is None:
+            # case 1: no active range, start active range.
+            cstart, cend = start, end
+        elif cend >= start - 1:
+            # case 2: active range continues through this range
+            # extend active range
+            cend = max(cend, end)
+        else:
+            # case 3: active range ends before this range begins
+            # write new range out, then start new active range
+            res_append((cstart, cend))
+            cstart, cend = start, end
+    if cend is not None:
+        res_append((cstart, cend))
+    return res
+
+
+def parse_ranges(rmaps, covers, chunk=10000):
+    """Extract range information from read maps.
+
+    Parameters
+    ----------
+    rmaps : dict of dict
+        Read map data structure.
+    covers : dict of list
+        Subject coverage data structure.
+    chunk : int, optional
+        Auto-compress after adding this many new ranges.
+
+    Notes
+    -----
+    The `covers` data structure is a dictionary, where the key is a tuple of
+    (sample, subject), and the value is a list. The first element of the list
+    is an integer indicating how many new ranges have been added since last
```

Comment:
This looks like bad practice.  It feels very wrong to mix data types within a list, especially in a non interleaved way.  It will make all for loops more complicated, as evidenced by the use of the [1:] everywhere.  The chunking is a memory optimization, you can just as easily use len(cover) rather than cover[0] if you want to know when to run the merge step.  If you're intent on triggered based on number of items added rather than number of items kept, that count should be kept separate from the list itself- either creating an object that contains the list and the count, or creating two separate covers maps, covers containing the actual range lists and cover_counts keeping track of how many items have been added.

Reply:
I agree with your opinion as I also feel bad with this solution.

It is ugly but it is the fastest solution I could think of. In this context, performance is the most important thing. Initially I used two separate dictionaries -- one for ranges and the other for counter. However, the query of dictionary is inefficient when there are many keys in it, which is the case here. Therefore, that solution (independent queries of two dictionaries) turned out to be inefficient.

Your suggestion of `[items_added, [range, range, range]]` is a good idea. I just tested. Your solution and my current solution have very close performance (runtime and memory). Let me update the PR using this solution.

I once considered using `len(cover)`. However, the logic prohibiting it is that the ranges may still be very fragmental after merging. This is very common when the sequencing depth is low for some genomes. Therefore, the length of `cover` may never drop below a certain line, no matter we merge it or not. If we use `len(cover)` as the mechanism to trigger auto-merge, the threshold must not be too small, and we don't know how big it should be. The worst case is that the function will futily merge the ranges during every call and no merge actually happens.